### PR TITLE
DO NOT MERGE: Comments from Jon

### PR DESCRIPTION
**Pull request by joncinque**
This is really great work! I really like the flow of the full overview, followed by the real-world usage, and then the challenge.  You've had to fight through quite a bit to understand all this, but you definitely get it :sweat_smile: 

If you want, we can deploy a version of token-swap to devnet based on the current production code, since I'm not sure if everything is synced up unfortunately.

Anyway, take it or leave it, but here are some comments :grin:

### CODE REVIEW

File: content/token-program.md [hunk 1]
```diff
@@ -65,6 +65,8 @@ Under the hood, the `createMint` function is simply creating a transaction that
 1. Create a new account 
 2. Initialize a new Mint 
 
+JC nit: Maybe explain why this is important -- that if you do them in separate transactions, someone else could scoop your created account for their own mint!
```
Comment (jamesrp13) [inline content/token-program.md:68]: @joncinque Does adding this on line 104 after the code snippet fit what you're suggesting?

> When manually building the instructions to create a new token mint, make sure you add the instructions to create the account and initialize the mint to the same transaction. If you were to do each step in a separate transaction, it's theoretically possible for somebody else to take the account you create and initialize it for their own mint.

Comment (jamesrp13) [inline content/token-program.md:68]: @joncinque Does adding this on line 104 after the code snippet fit what you're suggesting?

> When manually building the instructions to create a new token mint, make sure you add the instructions to create the account and initialize the mint to the same transaction. If you were to do each step in a separate transaction, it's theoretically possible for somebody else to take the account you create and initialize it for their own mint.

Comment (joncinque) [reply content/token-program.md:68]: Perfect!

File: content/token-program.md [hunk 2]
```diff
@@ -258,8 +260,10 @@ The `mintTo` function returns a `TransactionSignature` that can be viewed on the
 - `destination` the token account that tokens will be minted to
 - `authority` the account authorized to mint tokens
 - `amount` the amount of tokens to mint - remember to account for the decimals of the `mint`!
+JC nit: can make this even clearer to say that this is the raw token amount, outside of decimals
 
 It's not uncommon to update the mint authority on a token mint to null after the tokens have been minted. This would set a maximum supply and ensure no tokens can be minted in the future. Conversely, minting authority could be granted to a program so tokens could be automatically minted at regular intervals or according to programmable conditions.
+JC nit: an example here could be really useful!
```
Comment (jamesrp13) [inline content/token-program.md:266]: Note to self to maybe come back to this

Comment (jamesrp13) [inline content/token-program.md:266]: Note to self to maybe come back to this

File: content/token-program.md [hunk 3]
```diff
@@ -340,6 +345,8 @@ async function buildTransferTransaction(
 }
 ```
 
+JC: in the name of encouraging associated token accounts, this could take in the destination wallet address and use `getOrCreateAssociatedTokenAccount` to get the destination token account
```
Comment (jamesrp13) [inline content/token-program.md:348]: @joncinque this comment is in reference to the code snippet above, correct? If so, this snippet is only meant to show the implementation of the `spl-token` library's `transfer` function, which from what I can tell assumes that the address being passed into the function is already the correct token account

Comment (jamesrp13) [inline content/token-program.md:348]: @joncinque this comment is in reference to the code snippet above, correct? If so, this snippet is only meant to show the implementation of the `spl-token` library's `transfer` function, which from what I can tell assumes that the address being passed into the function is already the correct token account

File: content/token-program.md [hunk 4]
```diff
@@ -391,9 +398,14 @@ async function buildBurnTransaction(
 }
 ```
 
+JC:
+## Approve / Revoke Delegate
```
Comment (jamesrp13) [inline content/token-program.md:402]: Added this to our project board and will get to it ASAP

File: content/token-swap.md [hunk 5]
```diff
@@ -12,6 +12,7 @@
 # TL;DR
 
 - The **token swap program** is an SPL contract deployed to Mainnet and Devnet available for use by developers and protocols.
+JC: they shouldn't use those programs and instead deploy their own -- we don't support it on any of the networks
```
Comment (jamesrp13) [inline content/token-swap.md:15]: @joncinque sorry I'm not sure what "those programs" refers to here

File: content/token-program.md (hunk 3 above)
Comment (joncinque) [reply content/token-program.md:348]: That's correct.  The comment was a small suggestion in case you wanted to do a bit more with the example, since it's just calling `createTransferInstruction` otherwise

File: content/token-swap.md (hunk 5 above)
Comment (joncinque) [reply content/token-swap.md:15]: Sorry that wasn't clear -- I meant the token-swap program deployed to devnet and mainnet.  I mention it a bit later, but those are very old deployments that probably have loads of bugs, and aren't even managed by us.  If you wanted, we could deploy an "official" one to devnet that we keep up to date for people to test with.

Comment (jamesrp13) [reply content/token-swap.md:15]: Shoot...this is making me feel like I'm missing something crucial here 😬 

So does that render everything we're doing in the lesson incorrect then?

File: content/token-swap.md [hunk 6]
```diff
@@ -41,9 +43,12 @@ Creating swap pools with the SPL token swap program really showcases the account
...
 1. **Token swap state account** - holds information about the swap pool
 2. **Swap pool authority** - the account used to sign transactions on behalf of the swap program
+JC: you explain it later, but you can specify that this is a program-derived address based on the token swap state account
 3. **Token accounts for Token A and Token B** - associated token accounts for each of the tokens and the swap pool
+JC: "associated token account" is also a loaded term, so maybe strike the word since it isn't necessary?
 4. **Pool token mint** - the mint for the swap pool's LP-token
 5. **Pool token account** - the token account for the pool token mint
+JC: This is actually the pool token account that receives the first minting of pool tokens, representing the value inside of the token A and B accounts
```
Comment (jamesrp13) [inline content/token-swap.md:51]: @joncinque To clarify, you mean that when the account is first created, some of the LP tokens are minted to this account, but all subsequent LP-tokens are minted directly to the person depositing liquidity?

File: content/token-swap.md [hunk 7]
```diff
@@ -150,10 +155,14 @@ transaction.add(createTokenAccountPoolInstruction)
 transaction.add(initializeTokenAccountPoolInstruction)
 ```
 
+JC: feel free to use the creator's associated token account here!
+
 ### Pool Token Fee Account
 
 The pool token fee account is the token account that the fees for the token swaps are paid to. This account must be owned by a specific account defined in the swap program - that account has public key [HfoTxFR1Tm6kGmWgYWD6J7YHVy1UwqSULUGVLXkJqaKN](https://explorer.solana.com/address/HfoTxFR1Tm6kGmWgYWD6J7YHVy1UwqSULUGVLXkJqaKN?cluster=devnet).
 
+JC: if we encourage people to use their own program (or deploy or own), then it doesn't need to be owned by that address, it's a holdover from when Serum was running token-swap.
```
Comment (jamesrp13) [inline content/token-swap.md:164]: @joncinque The account address would be defined by whatever swap program they're using, correct? Or is there some deterministic way to know the account based on the token swap program's address?

File: content/token-swap.md (hunk 6 above)
Comment (jamesrp13) [reply content/token-swap.md:51]: Sweet I'll update some of the language to reflect that

File: content/token-swap.md (hunk 7 above)
Comment (joncinque) [reply content/token-swap.md:164]: It *can* be, if the program has constraints defined.  The previous deployment forced the fee owner to be a particular address, but that isn't the default behavior.  You can read more about constraints at https://github.com/solana-labs/solana-program-library/tree/master/token-swap#building -- either way, this is *not* typically deterministic

Comment (jamesrp13) [reply content/token-swap.md:164]: Got it
